returnn/forward.py

```diff
@@ -207,3 +216,188 @@ def hash(cls, kwargs):
+        :param device: RETURNN device, cpu or gpu
+        :param time_rqmt: job time requirement in hours
+        :param mem_rqmt: job memory requirement in GB
+        :param cpu_rqmt: job cpu requirement
+        """
+        self.returnn_config = returnn_config
+        self.model_checkpoint = model_checkpoint
+        self.returnn_python_exe = returnn_python_exe
+        self.returnn_root = returnn_root
+        self.log_verbosity = log_verbosity
+        self.device = device
+
+        self.out_returnn_config_file = self.output_path("returnn.config")
+        self.out_files = {output: self.output_path(output) for output in output_files}
+
+        self.rqmt = {
+            "gpu": 1 if device == "gpu" else 0,
+            "cpu": cpu_rqmt,
+            "mem": mem_rqmt,
+            "time": time_rqmt,
+        }
+
+    def tasks(self):
+        yield Task("create_files", mini_task=True)
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    def create_files(self):
+        """create files"""
+        config = self.create_returnn_config(
+            model_checkpoint=self.model_checkpoint,
+            returnn_config=self.returnn_config,
+            log_verbosity=self.log_verbosity,
+            device=self.device,
+        )
+        config.write(self.out_returnn_config_file.get_path())
+
+        cmd = [
+            self.returnn_python_exe.get_path(),
+            os.path.join(self.returnn_root.get_path(), "rnn.py"),
+            self.out_returnn_config_file.get_path(),
+        ]
+        util.create_executable("rnn.sh", cmd)
+
+        # check here if model actually exists
+        if self.model_checkpoint is not None:
+            assert os.path.exists(
+                _get_model_path(self.model_checkpoint).get_path()
+            ), f"Provided model checkpoint does not exists: {self.model_checkpoint}"
+
+    def run(self):
+        """run"""
+        # run everything in a TempDir as writing files can cause heavy load
+        with tempfile.TemporaryDirectory(prefix=gs.TMP_PREFIX) as tmp_dir:
+            print("using temp-dir: %s" % tmp_dir)
+            call = [
+                self.returnn_python_exe.get_path(),
+                os.path.join(self.returnn_root.get_path(), "rnn.py"),
+                self.out_returnn_config_file.get_path(),
+            ]
+
+            try:
+                env = os.environ.copy()
+                env["OMP_NUM_THREADS"] = str(self.rqmt["cpu"])
+                env["MKL_NUM_THREADS"] = str(self.rqmt["cpu"])
+                subprocess.check_call(call, cwd=tmp_dir, env=env)
+            except Exception:
+                print("Run crashed - copy temporary work folder as 'crash_dir'")
+                if os.path.exists("crash_dir"):
+                    shutil.rmtree("crash_dir")
+                shutil.copytree(tmp_dir, "crash_dir")
+                raise
+
+            # move logs
+            if os.path.exists(f"{tmp_dir}/returnn.log"):
```

Comment:
Note that I just copied the `ReturnnForwardJob` here, which does the same, i.e. has this hardcoded log name.

It's a bit tricky, definitely would add complexity here, but in principle, yes (to a certain degree).

So, do you want that extra complexity here?

Reply:
The file logging can be more customized by the user. E.g. I used to pipe it through some script which does some more aggressive buffering in memory in order to reduce file-system access and to speed it up. You can pipe it through any script, which performs any kind of post processing or filtering.

If the user intentionally uses a different name here, I would say it's the user's fault. But I don't just want to restrict the possibilities here because some user might potentially do wrong or buggy things otherwise.

I had a more generic idea though: After we moved the output files, maybe we can just remove all the remaining files from this temp directory to the work dir? This is even simpler and then the user can use anything. Is there any problem with this approach?

Reply:
> Is there any problem with this approach?
No, this should be fine

Reply:
Implemented that now.